File: Week 05/id_362/MaximalSquare.swift
import Foundation

extension Solution {
    public func maximalSquare(_ matrix: [[Character]]) -> Int {
        if matrix.isEmpty || matrix.first?.isEmpty ?? true { return 0 }
        var result = 0
        let m = matrix.count
        let n = matrix[0].count
        var dp = [[Int]](repeating: [Int](repeating: 0, count: n + 1), count: m + 1)
        for i in 1...m {
            for j in 1...n {
                if matrix[i - 1][j - 1] == "1" {
                    dp[i][j] = min(min(dp[i][j - 1], dp[i - 1][j - 1]), dp[i - 1][j]) + 1
                    result = max(result, dp[i][j])
                }
            }
        }
        return result * result
    }
}

File: Week 05/id_277/maximal-square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix: return 0

        result = 0
        rows = len(matrix)
        columns = len(matrix[0])
        dp = [[0] * (columns + 1) for _ in range(rows + 1)]
        for i in range(1,rows+1):
            for j in range(1,columns+1):
                if matrix[i-1][j-1] == '1':
                    dp[i][j] = min(dp[i-1][j],dp[i][j-1],dp[i-1][j-1]) + 1
                    result = max(result,dp[i][j] ** 2)
        return result